google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java

```diff
@@ -550,15 +550,40 @@ public GoogleAdsClient build() {
 
       // Provides the credentials to the primer to preemptively get these ready for usage.
       Primer.getInstance().ifPresent(p -> p.primeCredentialsAsync(getCredentials()));
-      // Proceeds with creating the client library instance.
-      TransportChannelProvider transportChannelProvider = getTransportChannelProvider();
-      if (transportChannelProvider.needsHeaders()) {
-        transportChannelProvider = transportChannelProvider.withHeaders(getHeaders());
-      }
-      if (transportChannelProvider.needsEndpoint()) {
-        transportChannelProvider = transportChannelProvider.withEndpoint(getEndpoint());
+      // Checks if the TransportChannelProvider is of type LocalChannelProvider in which case we
+      // cannot cast to InstantiatingGrpcChannelProvider.
+      if (getTransportChannelProvider()
+          .getClass()
+          .toString()
+          .equals("class com.google.api.gax.grpc.testing.LocalChannelProvider")) {
+        TransportChannelProvider transportChannelProviderTesting =
+            getTransportChannelProvider().withHeaders(getHeaders());
+        if (transportChannelProviderTesting.needsEndpoint()) {
+          transportChannelProviderTesting =
+              transportChannelProviderTesting.withEndpoint(getEndpoint());
+        }
+        setTransportChannelProvider(transportChannelProviderTesting);
+      } else {
+        InstantiatingGrpcChannelProvider transportChannelProvider =
+            (InstantiatingGrpcChannelProvider) getTransportChannelProvider();
+        transportChannelProvider =
+            transportChannelProvider.toBuilder()
+                .setInterceptorProvider(
+                    () ->
+                        ImmutableList.of(
+                            new LoggingInterceptor(
+                                new RequestLogger(), getHeaders(), getEndpoint())))
+                .build();
+        // Proceeds with creating the client library instance.
+        transportChannelProvider =
```

Comment:
What if the user called setTransportChannelProvider(x) where x is InstantiatingGrpcChannelProvider?

I think the approach where we don't call setTransportChannelProvider seems more logical. Instead, we check if getTransportChannelProvider() == null and if so then we create an InstantiatingGrpcChannelProvider. Is there some reason not to do that here?

Reply:
You cannot check if getTransportChannelProvider() == null. There is null precondition on getTransportChannelProvider(), so if it is null, an error is thrown.

Reply:
You should be able to add @Nullable to the method signature if that's a problem.

Reply:
To which method signature? TransportChannelProvider? We don't own that class.